javatests/arcs/core/allocator/AllocatorTestBase.kt

```diff
@@ -173,6 +203,213 @@ open class AllocatorTestBase {
+    extraParticle: A<Plan.Particle>
+  ) {
+    val registry = hostRegistry()
+    val newParticle = extraParticle()
+
+    // Precondition: extraParticle is *not* hosted by an ArcHost
+    assertThat(
+      registry.availableArcHosts().all { it.isHostForParticle(newParticle) == false }
+    ).isTrue()
+
+    // Invariant: starting a plan with an unhosted particle will throw an exception.
+    val allocator = allocator(registry)
+    val modifiedPlan = Plan.particleLens.mod(plan()) {
+      val list = it.toMutableList()
+      list.add(newParticle)
+      list
+    }
+    assertSuspendingThrows(ParticleNotFoundException::class) {
+      allocator.startArcForPlan(modifiedPlan)
+    }
+  }
+
+  suspend fun invariant_planWithOnly_mappedParticles_willResolve(
+    particles: A<List<ParticleInfo>>,
+    plan: T<List<Plan.Particle>, Plan>,
+    hostRegistry: T<List<ParticleRegistration>, HostRegistry>
+  ) {
+    // Setup: extract concrete values from input generators and transformations.
+    val theParticles = particles()
+    val registry = hostRegistry(theParticles.map { it.registration })
+    val thePlan = plan(theParticles.map { it.plan })
+
+    // Precondition: every particle in the plan is hosted by an ArcHost.
+    thePlan.particles.forEach {
+      assertThat(registry.availableArcHosts().any { host -> host.isHostForParticle(it) }).isTrue()
+    }
+
+    // Invariant: A plan that contains only hosted particles can be used to successfully start an Arc
+    val allocator = allocator(registry)
+    allocator.startArcForPlan(thePlan)
+  }
+
+  /**
+   * Test that adding an unmapped particle to a plan results in that plan being unable to be
+   * started.
+   */
+  @Test
+  fun allocator_verifyAdditionalUnknownParticleThrows() = runAllocatorTest {
+    val particle = Plan.Particle(
+      particleName = "Unknown Particle",
+      location = "unknown.Particle",
+      handles = emptyMap()
+    )
+
+    invariant_addUnmappedParticle_generatesError(
+      Value(PersonPlan),
+      Value(hostRegistry),
+      Value(particle)
+    )
+  }
+
+  /**
+   * Generate a random handleConnection map.
+   */
+  fun handleMapGenerator(s: Seed): A<Map<String, Plan.HandleConnection>> {
+    val storageKey = ChooseFromList(s, listOf("sk1", "sk2", "sk3", "sk4", "sk5", "sk6"))
+    return MapOf(
+      SequenceOf(listOf("a", "b", "c", "d", "e")),
+      HandleConnectionGenerator(
+        handle = HandleGenerator(
+          storageKey = CreatableStorageKeyGenerator(storageKey),
+          // TODO(shanestephens): Add a type generator
+          type = Value(PersonPlan.handles[0].type)
+        ),
+        mode = HandleModeFromType(s),
+        type = Value(PersonPlan.handles[0].type)
+      ),
+      ChooseFromList(s, listOf(1, 2, 3, 4, 5))
+    )
+  }
+
+  /**
+   * Test that adding a randomly generated unmapped particle will result in a plan being unable to
+   * be started.
+   */
+  @Test
+  open fun fuzz_addUnmappedParticle_generatesError() = runFuzzTest {
```

Comment:
this test is strange. It does not add any value to allocator_verifyAdditionalUnknownParticleThrows. I cannot imagine a situation where this would fail but allocator_verifyAdditionalUnknownParticleThrows would pass. I think we need some guidelines on when to use fuzz testing.

Reply:
This is here for demonstration purposes at the moment, it demonstrates a very simple addition of some bounded randomness. I would expect us to make some of the other parameters random too, eventually.

I want to pick you up on one thing, though: the whole point of this is picking up the bugs we didn't imagine could exist.

Reply:
I've modified this so that the hostRegistry is randomly generated from the list of particles.